volmigrate/backend/etcd2/etcd2.go

```diff
@@ -0,0 +1,117 @@
+package etcd2
+
+import (
+	"fmt"
+	"log"
+	"path"
+	"strconv"
+
+	"github.com/contiv/errored"
+	"github.com/contiv/volplugin/errors"
+	"github.com/contiv/volplugin/volmigrate/backend"
+	"github.com/coreos/etcd/client"
+	"golang.org/x/net/context"
+)
+
+func New(prefix string, etcdHosts []string) *Engine {
+	etcdCfg := client.Config{
+		Endpoints: etcdHosts,
+	}
+
+	etcdClient, err := client.New(etcdCfg)
+	if err != nil {
+		log.Fatalf("Failed to create etcd client: %s", err)
+	}
+
+	return &Engine{
+		etcdClient: client.NewKeysAPI(etcdClient),
+		prefix:     prefix,
+	}
+}
+
+// Engine is a migration engine for an etcd v2 datastore.
+type Engine struct {
+	etcdClient client.KeysAPI
+	prefix     string
+}
+
+// CurrentSchemaVersion returns the version of the last migration which was successfuly run.
+// If no previous migrations have been run, the schema version is 0.
+// If there's any error besides "key doesn't exist", execution is aborted.
+func (e *Engine) CurrentSchemaVersion() int64 {
+	resp, err := e.etcdClient.Get(context.Background(), path.Join(e.prefix, backend.SchemaVersionKey), nil)
+	if err != nil {
+		if err := errors.EtcdToErrored(err); err != nil && err == errors.NotExists {
+			return 0 // no key = schema version 0
+		}
+
+		log.Fatalf("Unexpected error when looking up schema version: %v\n", err)
+	}
+
+	i, err := strconv.Atoi(resp.Node.Value)
+	if err != nil {
+		log.Fatalf("Got back unexpected schema version data: %v\n", resp.Node.Value)
+	}
+
+	return int64(i)
+}
+
+// CreateDirectory creates a directory at the target path.
+func (e *Engine) CreateDirectory(target string) error {
+	fmt.Println("Creating directory: " + target)
+	if _, err := e.etcdClient.Set(context.Background(), path.Join(e.prefix, target), "", &client.SetOptions{Dir: true}); err != nil {
+		return errored.Errorf("Failed to create directory: %s", err)
```

Comment:
Hmm. I like the error type, but if we're not going to use it, that's fine. I would be extra careful to clearly describe what is happening (namely about migrations happening), though.